go/ocr2/decryptionplugin/decryption.go

```diff
@@ -0,0 +1,375 @@
+}
+
+// Report aggregates decryption shares from Observations to derive the plaintext.
+func (dp *decryptionPlugin) Report(ctx context.Context, ts types.ReportTimestamp, query types.Query, obs []types.AttributedObservation) (bool, types.Report, error) {
+	dp.logger.Debug("DecryptionReporting Report: start", commontypes.LogFields{
+		"epoch":         ts.Epoch,
+		"round":         ts.Round,
+		"nObservations": len(obs),
+	})
+
+	queryProto := &Query{}
+	if err := proto.Unmarshal(query, queryProto); err != nil {
+		return false, nil, fmt.Errorf("cannot unmarshal query: %w ", err)
+	}
+	ciphertexts := make(map[string]*tdh2easy.Ciphertext)
+	for _, request := range queryProto.DecryptionRequests {
+		ciphertext := &tdh2easy.Ciphertext{}
+		if err := ciphertext.UnmarshalVerify(request.Ciphertext, dp.publicKey); err != nil {
+			dp.logger.Error("DecryptionReporting Report: cannot unmarshal and verify the ciphertext, the leader is faulty", commontypes.LogFields{
+				"error":        err,
+				"ciphertextID": request.CiphertextID,
+			})
+			return false, nil, fmt.Errorf("cannot unmarshal and verify the ciphertext: %w", err)
+		}
+		ciphertexts[string(request.CiphertextID)] = ciphertext
+	}
+
+	fPlusOne := dp.genericConfig.F + 1
+	validDecryptionShares := make(map[string][]*tdh2.DecryptionShare)
+	for _, ob := range obs {
+		observationProto := &Observation{}
+		if err := proto.Unmarshal(ob.Observation, observationProto); err != nil {
+			dp.logger.Error("DecryptionReporting Report: cannot unmarshal observation, skipping it", commontypes.LogFields{
+				"error":    err,
+				"observer": ob.Observer,
+			})
+			continue
+		}
+
+		for _, decryptionShareWithID := range observationProto.DecryptionShares {
+			ciphertextID := string(decryptionShareWithID.CiphertextID)
+			ciphertext, ok := ciphertexts[ciphertextID]
+			if !ok {
+				dp.logger.Error("DecryptionReporting Report: there is not ciphertext in the query with matching id", commontypes.LogFields{
+					"ciphertextID": ciphertextID,
+					"observer":     ob.Observer,
+				})
+				continue
+			}
+
+			validDecryptionShare, err := dp.getValidDecryptionShare(ob.Observer,
+				ciphertext, decryptionShareWithID.DecryptionShare)
+			if err != nil {
+				dp.logger.Error("DecryptionReporting Report: invalid decryption share", commontypes.LogFields{
+					"error":        err,
+					"ciphertextID": ciphertextID,
+					"observer":     ob.Observer,
+				})
+				continue
+			}
+
+			validDecryptionShares[ciphertextID] = append(validDecryptionShares[ciphertextID], validDecryptionShare)
+			if len(validDecryptionShares[ciphertextID]) >= fPlusOne {
+				dp.logger.Trace("DecryptionReporting Report: we have already f+1 valid decryption shares", commontypes.LogFields{
+					"ciphertextID": ciphertextID,
+					"observer":     ob.Observer,
+				})
+				break
+			}
+		}
+	}
+
+	reportProto := Report{}
+	for id, decrShares := range validDecryptionShares {
+		ciphertext, ok := ciphertexts[id]
+		if !ok {
```

Comment:
seems to me like it is always `ok` here

Reply:
Yes, we have checked again [here](https://github.com/smartcontractkit/tdh2/blob/f0e8d0937917a332c1a3fb5c44798a6e76d24986/go/ocr2/decryptionplugin/decryption.go#LL238C1-L238C15).
It's a sanity check.